### PR TITLE
[core][android] Introduce OfflineManager.runPackDatabaseAutomatically(boolean) API

### PR DESCRIPTION
**Pull request by pozdnyakov**
Fixes mapbox/mapbox-gl-native-team#102

### CODE REVIEW

**Comment by chloekraw** [comment]
@pozdnyakov @tmpsantos thanks for this -- I like the idea a lot!

Right now I'm wondering whether the naming and documentation of these APIs could be further clarified.

I think one way we could add greater clarity is by explicitly referencing the SQLite `VACUUM` operation by name. The cache management APIs @tmpsantos added a few months ago already reference `VACUUM` in the inline documentation, e.g., https://github.com/mapbox/mapbox-gl-native/blob/753a640e9ffc3d9cfb6148eba078b3e9e57d9762/include/mbgl/storage/default_file_source.hpp#L215

File: include/mbgl/storage/default_file_source.hpp [hunk 1]
```diff
@@ -121,11 +121,6 @@ class DefaultFileSource : public FileSource {
      * Eviction works by removing the least-recently requested resources not also required
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: It's not clear from the documentation alone: is `VACUUM` triggered by the eviction invoked by this API?

File: include/mbgl/storage/default_file_source.hpp [hunk 2]
```diff
@@ -134,7 +129,7 @@ class DefaultFileSource : public FileSource {
      * executed on the database thread; it is the responsibility of the SDK bindings
      * to re-execute a user-provided callback on the main thread.
      */
-    void deleteOfflineRegion(OfflineRegion&&, std::function<void(std::exception_ptr)>, bool pack = true);
+    void deleteOfflineRegion(OfflineRegion&&, std::function<void(std::exception_ptr)>);
 
     /*
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: In what order are the APIs in this file listed? Currently, the order appears to be neither alphabetical nor logical...

if the order can be changed without too much hassle, let's do so in a separate pr

File: include/mbgl/storage/default_file_source.hpp [hunk 3]
```diff
@@ -195,6 +190,11 @@ class DefaultFileSource : public FileSource {
      */
     void packDatabase(std::function<void(std::exception_ptr)> callback);
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: Noting that this is a particularly confusing API name since we call offline regions "offline packs" on iOS 🤕 but I understand that this API name predates this PR.

Would we accept an update to the documentation like the following to clarify what the API does?

> Packs the existing database file into the minimum disk space required by invoking VACUUM on SQLite. If resources have been removed from the database, this operation would shrink the database size to reflect the remaining resources.

> When the operation is complete or encounters an error, the given callback will be executed on the database thread; it is the responsibility of the SDK bindings to re-execute a user-provided callback on the main thread.

(no change to the second paragraph)

File: include/mbgl/storage/default_file_source.hpp [hunk 4]
```diff
@@ -195,6 +190,11 @@ class DefaultFileSource : public FileSource {
...
 
+    /*
+     * Enables or disables auto database file packing after deleteing regions and clearing ambient cache.
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: Suggestion:

> Sets whether packing the database by invoking VACUUM on SQLite occurs automatically after an offline region is deleted (`deleteOfflineRegion()`) or the ambient cache is cleared (`clearAmbientCache()`).

> By default, packing is enabled. If disabled, the size of the database will not shrink after resources are removed unless `packDatabase()` is explicitly called.

File: include/mbgl/storage/default_file_source.hpp [hunk 5]
```diff
@@ -195,6 +190,11 @@ class DefaultFileSource : public FileSource {
...
+     */
+    void setAutopack(bool);
+
     /*
      * Forces revalidation of the ambient cache.
      *
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: If I'm understanding this PR correctly, we are refactoring `clearAmbientCache()` so that VACUUM is not always automatically run.

Therefore the documentation of `clearAmbientCache()` on lines 215-217 should be updated.

File: include/mbgl/storage/default_file_source.hpp [hunk 6]
```diff
@@ -195,6 +190,11 @@ class DefaultFileSource : public FileSource {
      */
     void packDatabase(std::function<void(std::exception_ptr)> callback);
 
+    /*
+     * Enables or disables auto database file packing after deleteing regions and clearing ambient cache.
+     */
+    void setAutopack(bool);
```
Comment (chloekraw) [inline include/mbgl/storage/default_file_source.hpp]: I think this API name is confusing. Naming this function `void runPackDatabaseAutomatically(bool)` would be a little clearer: the full reference to the related API is provided and it's obvious that if `true`, `packDatabase` is run. 

---

But why not implement this functionality so that `boolean runPackDatabase` is accepted as a variable in existing functions? e.g.:

> void deleteOfflineRegion(OfflineRegion&&, std::function<void(std::exception_ptr)>, bool runPackDatabase = true);

>  void clearAmbientCache(std::function<void (std::exception_ptr)>, bool runPackDatabase = true);

This way, the developer can decide with each function call whether to run `packDatabase()` automatically, and this new `"setAutopack"` API is not needed.

My understanding is that this wouldn't be a breaking change because if `runPackDatabase` is not passed in by the developer, it defaults to true which matches current behavior of both APIs. 

---

In the future, my preference would be to name all APIs around VACUUM/database defragmentation with those words explicitly. 

Unless `packDatabase()` does something additional to running `VACUUM`, what value do we provide the developer by wrapping a widely-known functionality (https://www.sqlitetutorial.net/sqlite-vacuum/) around this ambiguous word "pack"?

File: platform/android/src/offline/offline_manager.cpp [hunk 7]
```diff
@@ -211,6 +215,7 @@ void OfflineManager::registerNative(jni::JNIEnv& env) {
         METHOD(&OfflineManager::invalidateAmbientCache, "nativeInvalidateAmbientCache"),
         METHOD(&OfflineManager::clearAmbientCache, "nativeClearAmbientCache"),
         METHOD(&OfflineManager::setMaximumAmbientCacheSize, "nativeSetMaximumAmbientCacheSize"),
+        METHOD(&OfflineManager::setAutopack, "setAutopack"),
```
Comment (chloekraw) [inline platform/android/src/offline/offline_manager.cpp]: It seems based on the current naming convention, the name of this API should start with `native*`

Otherwise, why do some of these APIs start with `native*` and not others?

File: platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java [hunk 8]
```diff
@@ -569,7 +521,7 @@ public void run() {
   private native void getOfflineRegionStatus(OfflineRegionStatusCallback callback);
 
   @Keep
-  private native void deleteOfflineRegion(boolean pack, OfflineRegionDeleteCallback callback);
+  private native void deleteOfflineRegion(OfflineRegionDeleteCallback callback);
```
Comment (chloekraw) [inline platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/offline/OfflineRegion.java]: Same comment as [above](https://github.com/mapbox/mapbox-gl-native/pull/15967/files#r350525808), I think I'm missing an understanding of this proposal, since I think it's clearer to have `deleteOfflineRegion` and `nativeClearAmbientCache` accept a `boolean runPackDatabaseAutomatically`.

What I agree with is eliminating the `...AndSkipPackDatabase()` APIs.

File: include/mbgl/storage/default_file_source.hpp (hunk 3 above)
Comment (chloekraw) [reply include/mbgl/storage/default_file_source.hpp]: note: if this API does more than run vacuum, let's include in the first sentence ^

File: include/mbgl/storage/default_file_source.hpp (hunk 2 above)
Comment (pozdnyakov) [reply include/mbgl/storage/default_file_source.hpp]: I like the idea :+1: @tobrun @LukasPaczos wdyt?

File: platform/android/src/offline/offline_manager.cpp (hunk 7 above)
Comment (pozdnyakov) [reply platform/android/src/offline/offline_manager.cpp]: `native*` are the methods that are not exposed directly, i.e. require JAVA wrappers. `setAutopack` (`runPackDatabaseAutomatically`) is simple enough to be exposed as it is.

File: include/mbgl/storage/default_file_source.hpp (hunk 6 above)
Comment (pozdnyakov) [reply include/mbgl/storage/default_file_source.hpp]: > Unless packDatabase() does something additional to running VACUUM, what value do we provide the developer by wrapping a widely-known functionality (https://www.sqlitetutorial.net/sqlite-vacuum/) around this ambiguous word "pack"?

`VACUUM` is tied to sqlite, at some point we can switch to another DB provider. IMO it's better to keep API generic enough.

> But why not implement this functionality so that boolean runPackDatabase is accepted as a variable in existing functions?


This does not match well with the Android API: java does not support method parameters set by default.

Comment (pozdnyakov) [reply include/mbgl/storage/default_file_source.hpp]: Also, we do not run `VACUUM` - we run `PRAGMA incremental_vacuum` which provides a different behavior, so we'd better not to disclose the implementation internals in the API.

Comment (chloekraw) [reply include/mbgl/storage/default_file_source.hpp]: > VACUUM is tied to sqlite, at some point we can switch to another DB provider. IMO it's better to keep API generic enough.

> Also, we do not run VACUUM - we run PRAGMA incremental_vacuum which provides a different behavior, so we'd better not to disclose the implementation internals in the API.

Gotcha, makes sense to not reference VACUUM in the API name. Still, I think perhaps an alternative to "pack" should be considered for the future, unless there's a great reason why "pack" makes sense that I'm overlooking.

@pozdnyakov what do you think about specifying VACUUM / PRAGMA incremental_vacuum in the inline documentation though?

> This does not match well with the Android API: java does not support method parameters set by default.
 
😔 I guess our hands are tied here.

@pozdnyakov what do you think about changing the name from `setAutopack` to `runPackDatabaseAutomatically` then? (Also open to other suggestions.)

Finally -- is this function a global function for `OfflineManager`, so that if set, it would apply to every call to `deleteOfflineRegion` or `clearAmbientCache` that happens afterwards?

Comment (pozdnyakov) [reply include/mbgl/storage/default_file_source.hpp]: > @pozdnyakov what do you think about specifying VACUUM / PRAGMA incremental_vacuum in the inline documentation though?

We could mention it as possible implementation detail, however, it's better IMO to just describe what it does - shrink the database file size, and point out the potential performance issues.

> @pozdnyakov what do you think about changing the name from setAutopack to runPackDatabaseAutomatically then? (Also open to other suggestions.)

`runPackDatabaseAutomatically` looks much better, thanks for proposing it!

> Finally -- is this function a global function for OfflineManager, so that if set, it would apply to every call to deleteOfflineRegion or clearAmbientCache that happens afterwards?

yeah, it'll apply to all the following calls of deleteOfflineRegion or clearAmbientCache.

**Comment by pozdnyakov** [comment]
@chloekraw Thanks for your comments! They are considered in the latest patch set.

File: platform/default/src/mbgl/storage/offline_database.cpp [hunk 9]
```diff
@@ -705,7 +705,7 @@ std::exception_ptr OfflineDatabase::clearAmbientCache() try {
 
     resourceQuery.run();
 
-    vacuum();
+    if (autopack) vacuum();
```
Comment (tmpsantos) [inline platform/default/src/mbgl/storage/offline_database.cpp]: Optional: We could replace all the `vacuum()` calls with a protected `mayPack()` doing the `autopack == true` check and move the code inside `::vacuum` to `::pack`.

Comment (pozdnyakov) [reply platform/default/src/mbgl/storage/offline_database.cpp]: We've places where vacuum should be called unconditionally, e.g. at database creation in order to enable incremental vacuum

**Comment by pozdnyakov** [comment]
**Note for the change log:** Introduce `OfflineManager.runPackDatabaseAutomatically(boolean)` and remove the redundant `OfflineRegion.deleteAndSkipPackDatabase()` method.

**Comment by tobrun** [comment]
I will run point on integrating these changes in mapbox-gl-native-android
